zips/zip-0231.md

```diff
@@ -0,0 +1,164 @@
+> Created: 2024-04-26
+> License: MIT
+
+
+# Terminology
+
+{Edit this to reflect the key words that are actually used.}
+The key words "MUST", "REQUIRED", "MUST NOT", "SHOULD", and "MAY" in this
+document are to be interpreted as described in BCP 14 [#BCP14]_ when, and
+only when, they appear in all capitals.
+
+
+# Abstract
+
+TBD
+
+
+# Motivation
+
+In Zcash transaction versions v2-v5, each shielded output contains a ciphertext
+comprised of a 52-byte note plaintext, and a corresponding 512-byte memo field.
+Recipients can only decrypt the outputs sent to them, and thus can also only
+observe the memo fields included with the outputs they can decrypt.
+
+The shielded transaction protocol hides the sender(s) (that is, the addresses
+corresponding to the keys used to spend the input notes) from all of the
+recipients. For certain kinds of transactions, it is desirable to make one or
+more sender addresses available to one or more recipients (for example, a reply
+address), and in subsets of those transactions it is desirable to authenticate
+the sender addresses (to give the recipient a guarantee that the address is
+controlled by a sender of the transaction). These Authenticated Reply Addresses
+require zero-knowledge proofs, and for the Orchard protocol these proofs are too
+large to fit into a 512-byte memo field.
+
+Something about light client truncation (simplifies the security argument).
+
+In the light client protocol, a recipient need not download full transaction
+information if the compact transaction information indicates that they do not
+receive any memo in the transaction.
+
+
+# Requirements
+
+- Recipients can receive memo data that is greater than 512 bytes in length.
+- Multiple recipients, across any of the shielded pools, can be given the
+  capability to view the same memo data.
+- The exact number and exact lengths of distinct decryptable memos should not
+  be revealed, even to the transaction recipients, although an upper bound on
+  the total length of memo data that the observer does not have the capability
+  to view will be leaked.
+- A recipient can determine whether or not they have been given the capability
+  to view any memo solely by decrypting the note ciphertext.
+- Memo chunks within a transaction can be individually pruned without preventing
+  the transaction from being verified.
+  - As a consequence, a recipient with the capability to view a memo can prune
+    all of that memo's chunks.
+
+
+# Non-requirements
+
+- Recipients do not need to be able to receive multiple memos per note.
+
+
+# Specification
+
+## Memo encryption
+
+Sketch:
+
+In shielded bundles, the 512-byte memo field is replaced by `[flag_byte] || memo_key`.
+
+```
+STREAM: AEAD(key, counter, pt)
+
+[
+    (key_a, 0),
+    (key_b, 0),
+    (key_a, 1),
+    (key_c, 0),
+    (key_c, 1),
+    (key_a, 2),
+]
+```
+
+## Memo bundle format
+
+```rust
+enum MemoChunk {
+    Encrypted(AEAD(MemoChunk, memo_key)),
```

Comment:
We haven't yet decided the size of a memo chunk. In previous discussions we've assumed 512 bytes as a baseline. In today's discussion we briefly touched on the topic, noting that we may want to use a smaller chunk size (e.g. 256 bytes) to have greater flexibility (particularly given the consensus limit on memo bundle size), at the cost of more AEAD overhead and more information revealed about candidate memo lengths

Reply:
A minor variant here: if we used a 256 byte chunk size, then we could also require that the number of chunks be a multiple of two, with the fee being one marginal fee per chunk pair. This would have the effect that each action in a "standard" base-fee-only transaction would be able to to send 256 bytes to the recipient and record a 256-byte change chunk for themselves. I think that this would be a desirable feature.

Reply:
We decided on a 256-byte chunk size.